### DISCUSSION TITLE
Using KDL for config settings in Camping

### DISCUSSION THREAD

**Discussion by karloscarweber** [kdl-org/kdl#287] [Applications]
Hi Friends,

I'm working on [Camping](https://github.com/camping/camping)'s Upcoming 3.0 release and I want to use **KDL** to add config and database settings. Part of Camping's 3.0 release is to remove the hard ActiveRecord requirement and to use a plugin ([guidebook](https://github.com/karloscarweber/guidebook)) instead.

So, I wanna have a `config.kdl` for the database configuration, at least to start. Hopefully replacing all other settings spots with a `campguide.kdl` file in the project. This is what I have so far:

```
// config.kdl

database {
  development {
    adapter "sqlite3"
    database "dbcamping.db"
    host "localhost"
    pool 5
    timeout 5000
  }
}
```

Does this make sense? is this a good use of KDL or am I being obtuse?

**Comment by tabatkins** [comment]
Yeah, that seems great.

**Reply by zkat** [reply]
agreed. This seems like a great use for it! It's definitely a language designed for human editing of configuration files.

**Comment by karloscarweber** [comment]
Awesome!

I wanna show you what I have so far. My Config document. Just database stuff for now:

```kdl
// config.kdl

  database {
    environment "development" \
      adapter="sqlite3" \
      database="db/camping.db" \
      host="localhost" \
      pool=5 \
      timeout=5000
  }
```

In Ruby I'm looking through the db folder for any `.kdl` document and grabbing the first one. Then I'm parsing it. My plan is that each database node will have a number of environment sub nodes, and they will be used based on the environment that Camping is running in.

At first I had each listing a sub node of environment, but that didn't work the way I expected, which is actually good because it makes more sense that each setting of an environment is a property of that environment.

I think this works well. I'm going to try to add some error handling and tests.

Please let me know if I'm using kdl properly. I really like this.

**Comment by karloscarweber** [comment]
So I refined the settings definition and I'm really impressed at how it looks:
```kdl
  database {
    development adapter="sqlite3" database="db/camping.db" host="localhost" pool=5 timeout=5000
    production adapter="sqlite3" database="db/camping.db" host="localhost" pool=5 timeout=5000
  }
```

I think this is super great and I want to use kdl for all of Camping's settings.

**Reply by zkat** [reply]
You don't need kdl itself to do all the heavy lifting for you either, you could go as far as doing what I've seen some others do with defaults/shared configs:

```kdl
database {
  default adapter="sqlite3" database="db/camping.db" host="localhost" pool=5 timeout=5000
  development
  production adapter="postgres" database="$DATABASE_URL"
}
```

**Reply by karloscarweber** [reply]
Oh this is really smart. I have some defaults hard coded into the plugin but this is pretty cool and exposes the defaults really well.

I guess I would just do a merge after parsing the nodes to map the values properly. I'm gonna go try this now.

**Reply by karloscarweber** [reply]
Where is the `$DATABASE_URL` coming from in this instance? Would I just do a REGEX replace before parsing the kdl document?

**Reply by zkat** [reply]
presumably the environment. You can make `database` work however you want, or even do something like `database=(env)"$DATABASE_URL"` if you want to me extra clever.
